UserScrape/crawler.py

```diff
@@ -121,108 +135,145 @@ def login(self) -> CrawlResult:
-            if len(i.text) != 0:
-                recos.append(i.get_attribute('href').replace('https://www.youtube.com/watch?v=', ''))
-        self.get_video_features(source, recos)
-        return recos[0:branching]
+            personalized = 'Recommended for you' in i.text
+            if personalized:
+                personalized_counter += 1
+            # take the link and remove everything except for the id of the video that the link leads to
+            recommendation_id = i.get_attribute('href').replace(
+                'https://www.youtube.com/watch?v=', '')
+            recos[recommendation_id] = {
+                'id': recommendation_id, 'personalized': personalized}
+        # store the information about the current video plus the corresponding recommendations
+        self.get_video_features(source, recos, personalized_counter)
+        # return the recommendations 
+        return recos
+
+    def delete_last_video_from_history(self):
+        self.driver.get('https://www.youtube.com/feed/history')
+        # todo: maybe replace the full xpath because its ugly, but it works
+        last_video_delete_button = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.XPATH,
+                                            '/html/body/ytd-app/div/ytd-page-manager/ytd-browse/ytd-two-column-browse'
+                                            '-results-renderer/div[1]/ytd-section-list-renderer/div['
+                                            '2]/ytd-item-section-renderer[1]/div[3]/ytd-video-renderer/div['
+                                            '1]/div/div['
+                                            '1]/div/div/ytd-menu-renderer/div/ytd-button-renderer/a/yt-icon-button'
+                                            '/button'))
+        ).click()
+
+    def delete_history(self):
+        self.driver.get('https://www.youtube.com/feed/history')
+        history_delete_button = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.XPATH,
+                                            '/html/body/ytd-app/div/ytd-page-manager/ytd-browse/ytd-two-column-browse'
+                                            '-results-renderer/div['
+                                            '2]/ytd-browse-feed-actions-renderer/div/ytd-button-renderer['
+                                            '1]/a/paper-button/yt-formatted-string'))
+        ).click()
+
+        confirm_button = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.XPATH,
+                                            '/html/body/ytd-app/ytd-popup-container/paper-dialog/yt-confirm-dialog'
+                                            '-renderer/div[2]/div/yt-button-renderer['
+                                            '2]/a/paper-button/yt-formatted-string'))
+        ).click()
+
+    def _get_seconds(self, duration: str):
```

Comment:
I'm not happy with this helper function, there is probably a more dense, pythonic way to do this